### ISSUE TITLE
Thread crash causing endless "Thread will be restarted" loops

### ISSUE THREAD

**Issue by DanielWinks** [home-assistant/appdaemon#849]
I'm seeing the following errors in my logs, right before an endless "Thread will be restarted" loop begins. This seems to only occur with a specific app, and as far as I can see, that app isn't doing anything particularly "odd". App in question is [here](https://github.com/DanielWinks/appdaemon/blob/master/conf/apps/return_home.py). 

This is on AppDaemon 4.0.0 container. Previously also happened with 4.0.0beta2.

    /usr/local/lib/python3.8/site-packages/appdaemon/__main__.py:134: DeprecationWarning: Task.all_tasks() is deprecated, use asyncio.all_tasks() instead
    pending = asyncio.Task.all_tasks()
    /usr/local/lib/python3.8/site-packages/appdaemon/plugins/mqtt/mqttplugin.py:350: DeprecationWarning: The loop argument is deprecated since Python 3.8,
    and scheduled for removal in Python 3.10.
    await asyncio.wait_for(utils.run_in_executor(self, self.start_mqtt_service, first_time_service), 5.0, loop=self.loop)
    /usr/local/lib/python3.8/site-packages/aiohttp/connector.py:964: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
    hosts = await asyncio.shield(self._resolve_host(
    /usr/local/lib/python3.8/site-packages/aiohttp/locks.py:21: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
    self._event = asyncio.Event(loop=loop)
    /usr/local/lib/python3.8/asyncio/events.py:81: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
    self._context.run(self._callback, *self._args)
    /usr/local/lib/python3.8/site-packages/aiohttp/connector.py:964: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
    hosts = await asyncio.shield(self._resolve_host(
    /usr/local/lib/python3.8/site-packages/aiohttp/locks.py:21: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
    self._event = asyncio.Event(loop=loop)
    Exception in thread thread-8:
    Traceback (most recent call last):
    File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner
        self.run()
    File "/usr/local/lib/python3.8/threading.py", line 870, in run
        self._target(*self._args, **self._kwargs)
    File "/usr/local/lib/python3.8/site-packages/appdaemon/threading.py", line 759, in worker
        callback = "{}() in {}".format(funcref.__name__, name)
    AttributeError: 'NoneType' object has no attribute '__name__'
    2020-01-17 08:00:41.617856 CRITICAL AppDaemon: Thread thread-8 has died
    2020-01-17 08:00:41.618027 CRITICAL AppDaemon: Pinned apps were: ['return_home']
    2020-01-17 08:00:41.618125 CRITICAL AppDaemon: Thread will be restarted

I've put in an app to monitor my logs and restart AppDaemon entirely when this occurs, as it will never recover on its own.

What also interesting is this last occurred on my way to work. If you look at the app in question, it's listening for when a state change occurs, but only when `new='home'`. This shouldn't even trigger anything in said app when I leave.

Please take a look through that app and let me know if this is a bug in AppDaemon, or if perhaps there's something particular in that app which may be causing this and what I can change to avoid this situation.

Thanks.

**Comment by ReneTode** [comment]
1 thing i notice is that you nowhere set your handlers to None.
but you have "not None" in your if statement to cancel the handler.
devicetrackers can be very tricky. i have seen many cases where they turn on and off unexpectedly many times, so i think you might want to add logging to see if your trackers behave as expected.

using listeners inside listeners is very tricky. you really need to make 1000% sure that all listeners are canceled if you dont use the ONESHOT option.

i think in your case there are listeners not canceled (i am not sure but thats my gut feeling)
you can see if there are more then 1 listener active in the admin interface.

**Comment by DanielWinks** [comment]
@ReneTode Can you take a look at my main [config](https://github.com/DanielWinks/appdaemon/blob/master/conf/appdaemon.yaml)? The admin page loads, but every field in it is blank. Not sure if there's something there I'm missing, too. It was also blank across various AD v4 releases. I've tried both with and without the socketio option.

![Capture](https://user-images.githubusercontent.com/5339479/72623115-88784780-3912-11ea-969b-1082b7b9683e.PNG)


As for the app, I'll see if I can rework the listeners and maybe come up with a different method to handle what I'm doing (basically, waiting for a motion sensor to pick me up OR a set time to pass before playing an announcement, and ensuring it only happens once). TBH, I wrote most of that as part of my first app back on AD v3, so I'm sure I can come up with something better, especially if you think it's an issue with listeners inside listeners.

**Comment by ReneTode** [comment]
you have some old stuff in there, but that shouldnt bother. (api_port)
i never did put in transport, so that might be the case. (take it out to try)

for listeners like you want, we now got the oneshot option (see docs)
it listens for 1 change and then cancels the listener automaticly.

**Comment by DanielWinks** [comment]
> what browser do you use?

I've seen the same with Chrome and FF, on Windows 10 and Linux, as well as Chrome and Firefox on Android.

**Comment by ReneTode** [comment]
strange, are there errors in the console?
hmm, wait, i believe that @Odianosen25 had something simular. 
i think there was a small change right before the release.
hmm, now i did look back and see you got the 4.0 release

docker right? then odia is a better help anyway.

**Comment by DanielWinks** [comment]
No errors in the logs.

Didn't work on 4.0.0beta2 or now on the 4.0.0 release.

I'm running the pre-built container images from Dockerhub.

**Comment by DanielWinks** [comment]
Seems the admin page issue is related to HAProxy somehow... I tested it directly by forwarding a port into my container with kubectl and it works fine... I'll figure out what HAProxy needs for it.

I've re-written my app to cancel any possible callbacks before setting new ones, so there shouldn't be any listeners left 'dangling', I'll see if that helps. The "thread will be restarted" loop only happens every few days, randomly, so it'll be a bit to tell for sure.

**Comment by ReneTode** [comment]
the fact that it happens every few days, makes me even more think that its about extra listeners.
in that case slowly your resources are filling up.

**Comment by Odianosen25** [comment]
@DanielWinks,

That error means that one of your listeners, is not passing the right callback function when its executing. So a `None` is passed, when its expecting a function which leads to it dying. Please ensure your listeners are properly setup.

The other stuffs in the logs above, will be fixed in the next release.

Regards

**Comment by DanielWinks** [comment]
Alright, still working on troubleshooting this. Thread died again yesterday morning at 8:07AM. Home Assistant has zero events/state changes within 5 minutes of this, and nothing that should trigger any callbacks in this app within hours of when this thread died. The code for this app is linked in a earlier comment. The linked code is exactly what was running in AD during the last thread crash. Still on AD 4.0.0 using the Docker Hub provided container.

I've also updated the app prior to yesterday's crash, ensuring that any previous listeners and timers are cancelled prior to setting new ones, to make sure there's no 'dangling' ones out there. I've also added oneshot, immediate and timeout on the listen_state, to further help avoid this, to no avail. Even with all these changes, the thread has still crashed.

The main thing here is, why is this thread dying, for just this app, when there's no state changes at all in Home Assistant?

Per the admin console, this app has 2 state listeners and that's all, which is what's expected per the code in the app, and those state listeners have a properly defined callback signature, and it works 9 times out of 10. In fact, I don't think this "thread will be restarted" issue has ever occurred during the actual execution of this app. It's just random.

Basically the app is:
1. Listen for Mariah or I to return home
2. If Mariah, turn on the lights
3. If me, turn on lights and also wait 30 seconds or until a motion detector trips, play announcement
    A) Announcement needs to be either the instant motion is detected OR after 30 seconds (in case I'm fumbling with crap on the porch, etc)
    B) This timer + listen_state works fine. When it either the timer CB occurs OR the state CB,  it cancels the other (it, in fact, calls cancel on BOTH handles, so there shouldn't be a dangling listener)
    C) "Thread will be restarted" has never occurred during the execution of this either-or, so I don't think there's an issue here.

**Comment by ReneTode** [comment]
@DanielWinks you still dont set your handles back to None after canceling.
And did you aleady create an app to log all changes from the devicetrackers?

this part troubles me also:
```
        self.welcome_home_daniel_last_triggered = self.get_state(
            "script.welcome_home_daniel.last_triggered", namespace="persist",
        ) or self.set_state(
            "script.welcome_home_daniel.last_triggered",
            state=self.get_now_ts(),
            namespace="persist",
        )
```
i never use the return from set_state, but according to the docs set_state returns a dict.
but i dont see you using that var anywhere else, so it shouldnt give trouble.

did you create an custom namespace with the name "persist" in your appdaemon.yaml?

and entities shouldnt have more then 1 dot in their names.
so lines like this are wrong:
```
           self.set_state(
                "script.welcome_home_daniel.last_triggered",
                state=self.get_now_ts(),
                namespace="persist",
            )
```

somehow i get the feeling you mis out looking at the errorlog.

**Comment by DanielWinks** [comment]
I've handled the "listeners inside callbacks" thing that may have been causing this with some template sensors in Home Assistant. So while I'm not sure what was causing this, and you may want to look into maybe putting some sort of try/catch around this exception:
`File "/usr/local/lib/python3.8/site-packages/appdaemon/threading.py", line 759, in worker
    callback = "{}() in {}".format(funcref.__name__, name)`
But otherwise, this can be closed since it's not happening anymore.

**Comment by matteos1** [comment]
also for me
/usr/lib/python3.8/site-packages/appdaemon/__main__.py:134: DeprecationWarning: Task.all_tasks() is deprecated, use asyncio.all_tasks() instead
  pending = asyncio.Task.all_tasks()
/usr/lib/python3.8/site-packages/aiohttp/connector.py:964: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  hosts = await asyncio.shield(self._resolve_host(
/usr/lib/python3.8/site-packages/aiohttp/locks.py:21: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  self._event = asyncio.Event(loop=loop)
/usr/lib/python3.8/asyncio/events.py:81: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  self._context.run(self._callback, *self._args)
/config/appdaemon/apps/notifier/notifier_dispatch.py:49: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  cfg = yaml.load(ymlfile)
/usr/lib/python3.8/asyncio/events.py:81: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  self._context.run(self._callback, *self._args)
/usr/lib/python3.8/site-packages/aiohttp/connector.py:964: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  hosts = await asyncio.shield(self._resolve_host(
/usr/lib/python3.8/site-packages/aiohttp/locks.py:21: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  self._event = asyncio.Event(loop=loop)
Exception in thread Thread-7:
Traceback (most recent call last):
  File "/config/appdaemon/apps/notifier/alexa_manager.py", line 70, in worker
    self.volume_get(alexa_player, float(self.get_state(globals.get_arg(self.args, "default_restore_volume")))/100)
  File "/config/appdaemon/apps/notifier/alexa_manager.py", line 46, in volume_get
    self.dict_volumes[i] = self.get_state(i, attribute="volume_level", default=volume) if self.version >= '4.0.0' else volume
  File "/usr/lib/python3.8/site-packages/appdaemon/utils.py", line 191, in inner_sync_wrapper
    f = run_coroutine_threadsafe(self, coro(self, *args, **kwargs))
  File "/usr/lib/python3.8/site-packages/appdaemon/utils.py", line 285, in run_coroutine_threadsafe
    result = future.result(self.AD.internal_function_timeout)
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
  File "/usr/lib/python3.8/site-packages/appdaemon/adapi.py", line 1310, in get_state
    return await self.AD.state.get_state(
  File "/usr/lib/python3.8/site-packages/appdaemon/state.py", line 343, in get_state
    raise ValueError(
ValueError: Alexa_Manager: Querying a specific attribute is only possible for a single entity
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/config/appdaemon/apps/notifier/alexa_manager.py", line 124, in worker
    self.log(sys.exc_info()) 
NameError: name 'sys' is not defined
